# Cogs 118C, Spring 2020


## HW7

Name: Malachi Mabie

Student ID: A15737533

In [1]:
import numpy as np
import pandas as pd

# 1. Entropy w regard to visual stimulus action potential relations

In [2]:
# a) What is the entropy for the outcome of rolling a standard six-side die?
# P(x)= 1/6 for all x. Therefore,
entropy = -(6/6)*(np.log2(1/6))
print("entropy of unbiased die = ", entropy)

entropy of unbiased die =  2.584962500721156


In [3]:
# b) The frequency of each of the 4 nucleotides (A,C,G,T) in mammalian genomes is roughly:
#    P(A) = P(T) = .35; P(G) = P(C) = .15
# What is the entropy of a single nucelotide position in the genome?

entropy = -2*(.35)*np.log2(.35) - 2*(.15)*np.log2(.15)
print("entropy of single nucelotide position = ", entropy)

entropy of single nucelotide position =  1.8812908992306927


In [4]:
# c) Suppose I roll two standard dice, and let x1 and x2 be the outcomes of each roll.
#    What is the joint entropy, H(x1,x2)?

# freq per value is 1,2,3,4,5,6,5,4,3,2,1
entropy = (-2*(1/6)*np.log2(1/6) - 2*(2/6)*np.log2(2/6) - 2*(3/6)*np.log2(3/6) 
- 2*(4/6)*np.log2(4/6) - 2*(5/6)*np.log2(5/6) - np.log2(1))

print("joint entropy of two standard dice = ", entropy)

joint entropy of two standard dice =  4.136636511405688


#### d) Two rvs (X,Y) are indep if P(x,y)=P(x)P(y). Show that if x and y are independent then H(X,Y)=H(X)+H(Y) and therefore I(X;Y)=0.
    If P(x) and P(y) are independent, then their distributions are independent.
    Since entropy sums over all possible values of x and y, the calculation will be the same
    whether using separated or joint entropy calculation:

$$
H(X,Y) = \sum_{x,y} P(x,y) log_2{P(x,y)} = \sum_{x,y} P(x)P(y) log_2{(P(x)P(y))}
$$
$$
 = \sum_{x,y} P(x)P(y) log_2{P(x)} log_2{P(y)} = \sum_{x} P(x) log_2{P(x)} + \sum_{y} P(y) log_2{P(y)}
$$
$$
 = H(X) + H(Y)
$$

    Mutual information between two random variables is defined as:
$$
I(X;Y)=H(X)+H(Y)-H(X,Y)
$$

    Which with independence becomes
$$
H(X)+H(Y) - (H(X)+H(Y)) = 0
$$

#### e) What is the joint entropy for N dice rolls? Using this definition:
$$
H(x_1,x_2,...,x_N) = -\sum_{x_1} \sum_{x_2} ... \sum_{x_N} P(x_1, ... , x_N) log_2{P(x_1,...,x_N)}
$$

    Since the dice rolls are all independent random variables:
$$
P(x_1, ... , x_N) = P(x_1)P(x_2)...P(x_N)
$$

    Using our findings from part (d):
$$
H(x_1,x_2,...,x_N) = -\sum_{x_1} P(x_1) log_2{P(x_1)} - ...  -\sum_{x_N} P(x_N) log_2{P(x_N)} = H(x_1) + ... + H(x_N)
$$

    Incorporating the data in qustion (dice entropy), where N dice rolls is represented by X_N:
$$
H(X_N) = -N * log_2{1/6}
$$

In [5]:
# f) Download the table of single letter (monogram) and 2-letter (bigram) frequencies for English,
#    as well as a list of common English words, from source. Compute the entropy per letter for English text.
#    Make sure to normalize to get the probability of each letter.

bigrams = pd.read_csv("english_bigrams_1.txt",sep=" ", keep_default_na=False, na_values=['_'], header=None)
bigrams.columns = ["fragment", "frequency"]
bigrams.astype({"fragment":"str"})
display(bigrams.head())

monograms = pd.read_csv("english_monograms.txt",sep=" ", keep_default_na=False, na_values=['_'], header=None)
monograms.columns = ["fragment", "frequency"]
monograms.astype({"fragment":"str"})
display(monograms.head())

words = pd.read_csv("english_words.txt",sep=" ", keep_default_na=False, na_values=['_'], header=None)
words.columns = ["fragment", "frequency"]
words.astype({"fragment":"str"})
display(words.head())

# the letter combinations may be assumed to be independent.
data = monograms.append(bigrams.append(words))#.drop_duplicates(keep='last',inplace=False)

fragment  frequency
0       TH  116997844
1       HE  100689263
2       IN   87674002
3       ER   77134382
4       AN   69775179

fragment  frequency
0        E  529117365
1        T  390965105
2        A  374061888
3        O  326627740
4        I  320410057

fragment  frequency
0      THE   58784490
1       OF   25242547
2      AND   25188846
3       TO   24462483
4        A   22290078

In [6]:
# entropy calculations:
def entropy_calc(data):
    alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    entropies = []
    for letter in alphabet:
        l_freq = 0
        for fragment,frequency in data.itertuples(index=False):
            if(isinstance(fragment, str)):
                if(letter in fragment):
                    l_freq += frequency
            elif(letter=="A"):
                print(fragment,frequency)
        entropies.append(l_freq)
    entropies = pd.DataFrame(entropies)
    entropies = entropies / data.frequency.sum() # normalize
    entropies = -1 * entropies * np.log2(entropies) # calculate
    total_entropy = entropies.sum().loc[0] # total
    return(total_entropy)
monogram_entropy = entropy_calc(monograms)
print("entropy per letter, from monograms = ", monogram_entropy)

# print("entropy per letter from bigrams = ", entropy_calc(bigrams)) # 6.3151598262078465
# print("entropy per letter from words = ", entropy_calc(words)) # 9.061705244835611
# print("entropy per letter from all sources = ", entropy_calc(data)) # 5.913644745465774

entropy per letter, from monograms =  4.1846008280191365


In [7]:
# g) Now imagine a randomly generated “text” with these letter frequencies.
#    That is, assume that every letter is independent of the other letters.
#    What is the entropy of the distribution of 2-letter bigrams in this random text?

# using the given monogram frequencies, and under the independent combination suggestion:
bigram_from_monogram_entropy = 2 * monogram_entropy # custom distribution
print("entropy of the distribution of random text bigrams = ", bigram_from_monogram_entropy)

# # Because each letter in the bigram is independent, we can treat this like the dice problem.
# indep_bigram_entropy = -2 * np.log2(1/26) # uniform distribution
# print("entropy of a randomly generated bigram = ", indep_bigram_entropy) # 9.400879436282183


entropy of the distribution of random text bigrams =  8.369201656038273


In [8]:
# h) If you analyzed a randomly generated “text” with these letter frequencies,
#    what would be the probability of the letter combinations “QU” and “TH” ?
#    How about “HT” and “UQ”?  Compare these with the actual frequencies in English.

normal = monograms["frequency"].sum()

letter_combo_prob = pd.DataFrame(columns=["Letter combo", "Probability from letter freq", "Probability from bigram freq"],
                                index=[0,1,2,3])

letter_combo_prob.loc[0] =["QU", "{:.8f}".format((monograms[monograms["fragment"]=="Q"]["frequency"].values[0]
                                             * monograms[monograms["fragment"]=="U"]["frequency"].values[0]) / normal**2),
                     "{:.8f}".format(bigrams[bigrams["fragment"]=="QU"]["frequency"].values[0] / bigrams["frequency"].sum())]

letter_combo_prob.loc[1] =["TH", "{:.8f}".format((monograms[monograms["fragment"]=="T"]["frequency"].values[0] 
                                             * monograms[monograms["fragment"]=="H"]["frequency"].values[0]) / normal**2),
                    "{:.8f}".format(bigrams[bigrams["fragment"]=="TH"]["frequency"].values[0] / bigrams["frequency"].sum())]

letter_combo_prob.loc[2] =["HT", "{:.8f}".format((monograms[monograms["fragment"]=="T"]["frequency"].values[0]
                                             * monograms[monograms["fragment"]=="H"]["frequency"].values[0]) / normal**2),
                     "{:.8f}".format(bigrams[bigrams["fragment"]=="HT"]["frequency"].values[0] / bigrams["frequency"].sum())]

letter_combo_prob.loc[3] =["UQ", "{:.8f}".format((monograms[monograms["fragment"]=="Q"]["frequency"].values[0]
                                             * monograms[monograms["fragment"]=="U"]["frequency"].values[0]) / normal**2),
                      "{:.8f}".format(bigrams[bigrams["fragment"]=="UQ"]["frequency"].values[0] / bigrams["frequency"].sum())]

display(letter_combo_prob)


Letter combo Probability from letter freq Probability from bigram freq
0           QU                   0.00077773                   0.00096422
1           TH                   0.12349581                   0.02705698
2           HT                   0.12349581                   0.00193138
3           UQ                   0.00077773                   0.00000541

In [9]:
# i) Using the actual frequency of 2-letter bigrams in English, compute the entropy per bigram.
#    Compare this with your theoretical prediction (1g, above): Is it larger or smaller, and why?
print("entropy per letter from bigrams:", entropy_calc(bigrams))
print("entropy from 1g  prediction:\t", bigram_from_monogram_entropy)

entropy per letter from bigrams: 6.3151598262078465
entropy from 1g  prediction:	 8.369201656038273


    The actual entropy is less than the predicted entropy because the prediction supposes independence, 
    while in actuality there exists structure (grammar, vocab, usage norms) that reduces entropy.

#### j) Combine the previous results to compute the mutual information between the first and second letters in an English bigram.
$$
I(A;B) = H(A) + H(B) - H(A,B) = 2H(A) - H(A,B)
$$
$$
 = (8.369 - 6.315) = 2.054
$$


In [10]:
# k) Compute the actual and predicted frequencies of the words “NEURON” and “UCSD,”
#    and compare these with the prediction based on the independent letters model.
normal = words["frequency"].sum()

word_freqs = pd.DataFrame(columns=["Actual Frequency", "Predicted Frequency"],index=["NEURON","UCSD"])
word_freqs.loc["NEURON"]["Actual Frequency"] = words[words["fragment"]=="NEURON"]["frequency"].values[0]
word_freqs.loc["UCSD"]["Actual Frequency"] = words[words["fragment"]=="UCSD"]["frequency"].values[0]

def letter_prob(word):
    s = monograms["frequency"].sum()
    prob = 1
    for letter in word:
        prob *= monograms[monograms["fragment"]==letter]["frequency"].values[0]/s
    return prob 

word_freqs.loc["NEURON"]["Predicted Frequency"] = (normal*letter_prob("NEURON")).round()
word_freqs.loc["UCSD"]["Predicted Frequency"] = (normal*letter_prob("UCSD")).round()

display(word_freqs)

Actual Frequency Predicted Frequency
NEURON              858                  72
UCSD                422                2013

#### What is the log likelihood ratio, where h0 is the independent letters model and h1 is the model using the actual frequencies of English words?
$$
log{[P(w | h_1) / P(w | h_0)]} = log{P(w|h_1)} - log{P(w|h_0)}
$$


In [11]:
# calculating this ratio for the words given:
word_freqs["Likelihood Ratio"] = ["{:.2f}".format(np.log(858/72)), "{:.2f}".format(np.log(422/2013))]
display(word_freqs)

Actual Frequency Predicted Frequency Likelihood Ratio
NEURON              858                  72             2.48
UCSD                422                2013            -1.56

# 2. Poisson Spiking problem

Consider a (very simplified) Poisson spiking model of a retinal ganglion cell. The cell produces spikes at rate r_1=4 Hz when stimulus 1 is presented (e.g. a bright spot), and it spikes at rate r_2=1 Hz when stimulus 2 (a dark spot) is presented. Let X be the random variable describing the stimulus, and P(x=1)=P(x=2)=1/2. Let n{0,1} be the number of spikes emitted by the cell during a time interval of duration dt =10 ms following stimulus presentation. In this problem we will neglect the small probability of >1 spike occurring in a given time bin, so that
$$P(n=1|x=1)=r_1 dt=0.04$$
$$P(n=1|x=2)=r_2 dt=0.01$$
$$P(n=0|x)=1-P(n=1|x)$$

#### a) Compute P(n = 1). This is the marginal probability for n.
$$P(n)=P(n|x=1)P(x=1)+P(n|x=2)P(x=2)=\frac{1}{2}[P(n|x=1)+P(n|x=2)]$$

In [144]:
# a)
p = .5 # P(x)
x1 = 4*.01 # P(1|x1)
x2 = 1*.01 # P(1|x2)
n = (x1+x2)*p # P(n)
print("Marginal Probability P(n) = ", n)

Marginal Probability P(n) =  0.025


In [145]:
# b) Define the joint distribution, P(x,n)=P(n|x)P(x). Make a 2x2 table showing the joint probability.
_x1 = 1 - x1 # P(0|x1)
_x2 = 1 - x2 # P(0|x2)
_n = 1 - n
joint_dist = pd.DataFrame(columns=["x=1", "x=2"],index=["n=0","n=1"])
joint_dist.loc["n=0"] = [_x1, _x2]
joint_dist.loc["n=1"] = [x1, x2]
joint_dist = joint_dist*p

display(joint_dist)

x=1    x=2
n=0  0.48  0.495
n=1  0.02  0.005

In [146]:
# c) What are H(x), H(n), and H(x,n)?

Hx = -np.log2(p)
Hn = -n*np.log2(n) - _n*np.log2(_n)
Hxn = joint_dist.applymap(np.log2)
Hxn = Hxn * -joint_dist
Hxn = Hxn_12.sum().sum()
entropy = pd.DataFrame(columns=["H(x)", "H(n)", "H(x,n)"])
entropy.loc[0] = [Hx,Hn,Hxn]
display(entropy)

H(x)      H(n)    H(x,n)
0   1.0  0.168661  1.186543

In [149]:
# d) What is the mutual information I(x;n)?
#    (This is the “number of bits of information” that the neuron’s spike rate carries about the stimulus.)
mutual_info = (entropy["H(x)"]+entropy["H(n)"]-entropy["H(x,n)"])[0]
print("mutual information I(x;n): ", mutual_info)

mutual information I(x;n):  -0.017881730992492573


In [154]:
# e) time bin split in half.
#    There are four possible outcomes: n12=[0 0] (no spikes), n12=[0 1] or n12=[1 0] (one spike), or n12=[1 1] (2 spikes).
#    What is the joint distribution, P(x,n_12) and corresponding entropy, H(x,n_12)?
x1 = 4*.005
x2 = 1*.005
_x1=1-x1**2
_x2=1-x2**2
__x1 = 1 - 2*x1
__x2 = 1 - 2*x2

joint_dist = pd.DataFrame(columns=["x=1", "x=2"],index=["n=00","n=01","n=10","n=11"])
joint_dist.loc["n=00"] = [__x1, __x2]
joint_dist.loc["n=01"] = [x1*_x1, x2*_x2]
joint_dist.loc["n=10"] = joint_dist.loc["n=01"]
joint_dist.loc["n=11"] = [x1**2, x2**2]
joint_dist = joint_dist/joint_dist.sum()
joint_dist = joint_dist*p
display(joint_dist)

Hxn_12 = joint_dist.applymap(np.log2)
Hxn_12 = Hxn_12 * -joint_dist
Hxn_12 = Hxn_12.sum().sum()
print("H(x,n_12) = ", Hxn_12)


x=1          x=2
n=00     0.479816     0.494988
n=01   0.00999216   0.00249988
n=10   0.00999216   0.00249988
n=11  0.000199923  1.24997e-05

H(x,n_12) =  1.18919555082807


In [155]:
# f) Compute the marginal, P(n_12), and use this to find H(n_12).

Pn_12 = 1 - joint_dist.loc["n=00"].sum()
print("P(n_12) = ", Pn_12)
_Pn_12 = 1 - Pn_12

Hn_12 = -Pn_12*np.log2(Pn_12) - _Pn_12*np.log2(_Pn_12)
print("H(n_12) = ", Hn_12)

P(n_12) =  0.025196500195077687
H(n_12) =  0.1696983742860639


In [156]:
# g) Using these results, compute the mutual information I(x;n12).
mutual_info_12 = Hx + Hn_12 - Hxn_12
print("mutual information I(x;n_12): ", mutual_info_12)

mutual information I(x;n_12):  -0.019497176542006223


In [157]:
print("difference between mutual infos = ",mutual_info - mutual_info_12)

difference between mutual infos =  0.00161544554951365


#### h) Notice that I(x;n_12) ~= I(x;n). Comment on whether this neuron contains information about the stimulus in the timing of spikes (a “temporal code”) or only in the number of spikes (a “rate code”).
Because the mutual information was basically the same, we can discern that whether the spike occured slightly earlier or later didn't matter, but rather the rate coding of spikes coming in.
It is also important to note that the refresh rate for both signals is too slow for two of the same signal to occur in the time window.